lib/cdo/contact_rollups.rb

```diff
@@ -0,0 +1,576 @@
+    # Create it based on master contact_rollups table. Create it every time to keep up with schema changes in contact_rollups.
+
+    # We deliberately use Sequel rather than ActiveRecord for DB access here. If we are running in a test, the test framework
+    # runs the test inside a transaction on the ActiveRecord connection in order to be able to roll back the transaction
+    # and discard test changes at the end of the test. CREATE TABLE on that connection would cause an implicit commit
+    # of the transaction and break tests. So we use a different connection to run schema modification commands.
+    PEGASUS_REPORTING_DB.run "DROP TABLE IF EXISTS #{DEST_TABLE_NAME}"
+    PEGASUS_REPORTING_DB.run "CREATE TABLE #{DEST_TABLE_NAME} LIKE #{TEMPLATE_TABLE_NAME}"
+
+    log_completion(start)
+  end
+
+  def self.log(s)
+    puts s unless Rails.env.test?
+    CDO.log.info s
+  end
+
+  def self.log_completion(start)
+    log "   Completed in #{Time.now - start} seconds"
+  end
+
+  def self.insert_from_dashboard_contacts
+    start = Time.now
+    log "Inserting teacher contacts and IP geo data from dashboard.users"
+    ActiveRecord::Base.connection.execute "
+    INSERT INTO #{PEGASUS_DB_NAME}.#{DEST_TABLE_NAME} (email, name, dashboard_user_id, roles, city, state, postal_code, country)
+    SELECT email COLLATE utf8_general_ci, name, users.id, 'Teacher', city, state, postal_code, country FROM #{DASHBOARD_DB_NAME}.users AS users
+    LEFT OUTER JOIN #{DASHBOARD_DB_NAME}.user_geos AS user_geos ON user_geos.user_id = users.id
+    WHERE users.user_type = 'teacher' AND LENGTH(email) > 0 -- AND user_geos.indexed_at IS NOT NULL -- TODO BUGBUG
+    ON DUPLICATE KEY UPDATE #{DEST_TABLE_NAME}.name = VALUES(name), #{DEST_TABLE_NAME}.dashboard_user_id = VALUES(dashboard_user_id)"
+    log_completion(start)
+  end
+
+  def self.insert_from_dashboard_pd_enrollments
+    start = Time.now
+    log "Inserting contacts from dashboard.pd_enrollments"
+    ActiveRecord::Base.connection.execute "
+    INSERT INTO #{PEGASUS_DB_NAME}.#{DEST_TABLE_NAME} (email, name)
+    SELECT email, name FROM #{DASHBOARD_DB_NAME}.pd_enrollments AS pd_enrollments
+    WHERE LENGTH(pd_enrollments.email) > 0
+    ON DUPLICATE KEY UPDATE name = #{DEST_TABLE_NAME}.name"
+    log_completion(start)
+  end
+
+  def self.insert_from_pegasus_contacts
+    start = Time.now
+    log "Inserting contacts from pegasus.contacts"
+    ActiveRecord::Base.connection.execute "
+    INSERT INTO #{PEGASUS_DB_NAME}.#{DEST_TABLE_NAME} (email, opted_out, name)
+    SELECT email, IF(unsubscribed_at IS null, null, true) AS opted_out, name FROM #{PEGASUS_DB_NAME}.contacts WHERE LENGTH(email) > 0
+    ON DUPLICATE KEY UPDATE #{DEST_TABLE_NAME}.name = VALUES(name)"
+    log_completion(start)
+  end
+
+  def self.insert_from_pegasus_forms
+    start = Time.now
+    log "Inserting contacts and IP geo data from pegasus.forms"
+    ActiveRecord::Base.connection.execute "
```

Comment:
Same as in the test above, this repeated `ActiveRecord::Base.connection.execute` is a bit cumbersome and would be more readable aliased in a method.